forum/qa-include/pages/ask.php

```diff
@@ -1,290 +1,302 @@
+
+	$permiterror=qa_user_maximum_permit_error('permit_post_q', QA_LIMIT_QUESTIONS);
+
+	if ($permiterror) {
+		$qa_content=qa_content_prepare();
+
+		// The 'approve', 'login', 'confirm', 'limit', 'userblock', 'ipblock' permission errors are reported to the user here
+		// The other option ('level') prevents the menu option being shown, in qa_content_prepare(...)
+
+		switch ($permiterror) {
+			case 'login':
+				$qa_content['error']=qa_insert_login_links(qa_lang_html('question/ask_must_login'), qa_request(), isset($followpostid) ? array('follow' => $followpostid) : null);
+				break;
+
+			case 'confirm':
+				$qa_content['error']=qa_insert_login_links(qa_lang_html('question/ask_must_confirm'), qa_request(), isset($followpostid) ? array('follow' => $followpostid) : null);
+				break;
+
+			case 'limit':
+				$qa_content['error']=qa_lang_html('question/ask_limit');
+				break;
+
+			case 'approve':
+				$qa_content['error']=qa_lang_html('question/ask_must_be_approved');
+				break;
+
+			default:
+				$qa_content['error']=qa_lang_html('users/no_permission');
+				break;
+		}
+
+		return $qa_content;
+	}
+
+
+//	Process input
+
+	$captchareason=qa_user_captcha_reason();
+
+	$in['title']=qa_get_post_title('title'); // allow title and tags to be posted by an external form
+	$in['extra']=qa_opt('extra_field_active') ? qa_post_text('extra') : null;
+	if (qa_using_tags())
+		$in['tags']=qa_get_tags_field_value('tags');
+
+	if (qa_clicked('doask')) {
+		require_once QA_INCLUDE_DIR.'app/post-create.php';
+		require_once QA_INCLUDE_DIR.'util/string.php';
+
+		$categoryids=array_keys(qa_category_path($categories, @$in['categoryid']));
+		$userlevel=qa_user_level_for_categories($categoryids);
+
+		$in['name']=qa_post_text('name');
+		$in['notify'] = strlen(qa_post_text('notify')) > 0;
+		$in['email']=qa_post_text('email');
+		$in['queued'] = qa_user_moderation_reason($userlevel) !== false;
+
+		qa_get_post_content('editor', 'content', $in['editor'], $in['content'], $in['format'], $in['text']);
+
+		$errors=array();
+
+		if (!qa_check_form_security_code('ask', qa_post_text('code')))
+			$errors['page']=qa_lang_html('misc/form_security_again');
+
+		else {
+			$filtermodules=qa_load_modules_with('filter', 'filter_question');
+			foreach ($filtermodules as $filtermodule) {
+				$oldin=$in;
+				$filtermodule->filter_question($in, $errors, null);
+				qa_update_post_text($in, $oldin);
+			}
+
+            $categoriesEnabled = qa_using_categories() && count($categories) > 0;
+            $noSubcategory = qa_opt('allow_no_sub_category');
+            $mainCategory = qa_post_text('category_1');
+
+            $subcategories = array_filter($categories, function ($category) use ($mainCategory) {
+                return $category['parentid'] === $mainCategory;
+            });
+            $subcategories = array_column($subcategories, 'categoryid');
+
+            if ($categoriesEnabled && !qa_opt('allow_no_category') && !isset($in['categoryid'])) {
```

Comment:
Czy te ify można wydzielić do osobnej funkcji? Ich rolą jest warunkowe przypisanie tłumaczenia dla błędu kategorii, więc IMO można to zapisać `$errors['categoryid'] = getCategoryErrors(/* neizbędne parametry dla ifów */)`, a już w funkcji niech każdy if zwraca po prostu wynik wywołania `qa_lang_html( .. )`.

Reply:
Można, nie zrobiłem tego bo nigdzie w tym pliku nie ma funkcji i wszystko leci w taki sposób. Wiem, to trochę błędne koło, bo robiąc tak nie poprawimy kodu. Tyle że jeśli chcemy poprawić kod to wypadałoby zrobić np. refactor całego tego pliku i jakoś sensownie go przepisać, podzielić, pozmieniać. A to z kolei zajmie czas, a z czasem wiemy jak jest.

Reply:
> Tyle że jeśli chcemy poprawić kod to wypadałoby zrobić np. refactor całego tego pliku i jakoś sensownie go przepisać, podzielić, pozmieniać.

I tak i nie. Jeśli dodaje się nowy kod lub poprawia istniejący, to można dany fragment od razu napisać lepiej niż to było do tej pory w danym miejscu. Nie zajmie to dużo więcej czasu niż napisanie w "starym stylu", a na pewno będzie to jeden krok w stronę poprawy jakości kodu.

Reply:
Próbować można, ale wydaje mi się, że jednak znacznie lepiej jest przepisać jakąś całość, wtedy to jest spójne i ogarnięte. A tak mamy straszną mieszankę w jednym małym fragmencie. Jak chcesz to mogę coś z tym porobić jeszcze.

Reply:
Nie upieram się. Ale jeśli widzę, że coś można tanim kosztem poprawić, to przy okazji to robię. Decyzję zostawiam Tobie. ;)